metaseq/scripts/reshard_fsdp.py

```diff
@@ -0,0 +1,235 @@
+        :param output_shard_name: A string pattern specifying the path names of the output shards.
+            Shard indices can be included in the path names if the pattern includes `{i}`.
+            (e.g. "checkpoints/2.7B/reshard/reshard-model_part-0-shard{i}.pt").
+        :param num_output_shards: The number of output shards.
+        :param unflatten_weights: Specifies whether the weights in the input shards should be
+            unflattened. This option is only applicable when the number of output shards is 1.
+        :param skip_optimizer_state: Specifies whether to skip the optimizer states from the input shards.
+    """
+    files = glob(input_glob_pattern)
+    if len(files) == 0:
+        raise ValueError("The glob pattern doesn't match any sharded checkpoints.")
+    files = sorted(files, key=lambda x: list(map(int, re.findall(r"\d+", x))))
+    logger.info(f"Found {len(files)} sharded checkpoints ({files[0]} to {files[-1]})")
+
+    logger.info("Loading all sharded checkpoints to CPU")
+    shard_state_dicts = [torch.load(path, torch.device("cpu")) for path in files]
+
+    resharded_state_dicts = reshard_fsdp_state_dicts(
+        shard_state_dicts,
+        num_output_shards=num_output_shards,
+        unflatten_weights=unflatten_weights,
+        skip_optimizer_state=skip_optimizer_state,
+    )
+    for shard_idx, state_dict in enumerate(resharded_state_dicts):
+        output_file = output_shard_name.format(i=shard_idx)
+        os.makedirs(os.path.dirname(output_file), exist_ok=True)
+        logger.info(f"Writing a resharded state dict to {output_file}")
+        torch.save(state_dict, output_file)
+
+
+def reshard_fsdp_state_dicts(
+    shard_state_dicts: List[Dict[str, Any]],
+    num_output_shards: int = 1,
+    unflatten_weights: bool = True,
+    skip_optimizer_state: bool = False,
+) -> List[Dict[str, Any]]:
+    logger.info(f"Resharding state dicts from {len(shard_state_dicts)} shards")
+    # Unshard model weights
+    resharded_state_dicts = [{} for _ in range(num_output_shards)]
+    resharded_weights, resharded_metadata = reshard_fsdp_model_weights(
+        [state["model"] for state in shard_state_dicts],
+        [state["shard_metadata"] for state in shard_state_dicts],
+        num_output_shards,
+        unflatten_weights=unflatten_weights,
+    )
+    for shard_idx, (weight, metadata) in enumerate(
+        zip(resharded_weights, resharded_metadata)
+    ):
+        resharded_state_dicts[shard_idx]["model"] = weight
+        resharded_state_dicts[shard_idx]["shard_metadata"] = metadata
+
+    # Unshard last optimizer state
+    if not skip_optimizer_state and "last_optimizer_state" in shard_state_dicts[0]:
+        reshared_optim_states = reshard_fsdp_optim_state(
+            [state["last_optimizer_state"] for state in shard_state_dicts],
+            num_output_shards,
+        )
+        # TODO: Support optimizer state unpadding
```

Comment:
Will script work without this TODO? Which use-case is missing?

Reply:
This part is similar to the existing [reshard_mp.py](https://github.com/facebookresearch/metaseq/blob/5ca6005552b7a049f071754393d5b86185e0db6a/metaseq/scripts/reshard_mp.py#L141) script where no unpadding is applied. The other script [checkpoint_utils.py](https://github.com/facebookresearch/metaseq/blob/5ca6005552b7a049f071754393d5b86185e0db6a/metaseq/checkpoint_utils.py#L660) does have some unpadding logic, but I'm not entirely sure if it's correct since it tries to find the padding info for the optimizer states from that of the model parameters and presumably always get 0 padding. I need to double check if we actually save the padding info for the optimizer states in the existing checkpoints.

Reply:
So just to confirm my understanding, padding doesn't affect the final checkpoint files, it just increases storage size a bit? So we can live without this optimization for now?

Reply:
Ah, I've confirmed with Naman about the padding logic for the optimizer states. This part is now completed and tested.